### PR TITLE
[ux] Restyle notification widget #182

### LINKED ISSUE #182: [ux] Restyle notification widget
After improving ui of administration site we need to restyle notification widget.

### Tasks

- [ ] Create new design of the widget.
- [ ] Implement created design.

blocked by
https://github.com/openwisp/openwisp-utils/issues/188

### PR DESCRIPTION
**Pull request by niteshsinha17**
### Changes
Expected design: https://miro.com/app/board/o9J_lMAQQZk=/
1. Restyled notification widget.
2. Changed unsubscribe button.
3. Updated CI run commands because tests were failing.
4. Updated readme.
5. Manually tested current changes.

### Tasks

- [x] Create a design for the widget.
- [x] Implement the created design.

### Checklist

- [x] I have read the [contributing guidelines](http://openwisp.io/docs/developer/contributing.html#how-to-commit-your-changes-properly).
- [x] I have manually tested the proposed changes.
- [x] I have written new test cases to avoid regressions. (if necessary)
- [x] I have updated the documentation. (e.g. README.rst)

close #182

### CODE REVIEW

**Comment by niteshsinha17** [comment]
@nemesisdesign  and @atb00ker I have made some changes in ci because it was not installing the correct version of ``utils`` and ``users`` so tests were falling. But that was not an issue. 

Problem is that menu element is getting registered twice at the same position from ``users`` during the test. Please check test's output. Do you have any idea why it's happening? I am facing the same issue with ``controller``.

**Review by nemesisdesign — COMMENTED** [review]
@niteshsinha17 that's a special test which messes with the installed apps, please read the code of the test to understand what it does so you should figure out a solution.

File: openwisp_notifications/static/openwisp-notifications/images/icons/icon-bell.svg [hunk 1]
```diff
@@ -1,6 +1 @@
-<svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="bell" class="svg-inline--fa fa-bell fa-w-14"
-    role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512">
-    <path fill="white"
-        d="M224 512c35.32 0 63.97-28.65 63.97-64H160.03c0 35.35 28.65 64 63.97 64zm215.39-149.71c-19.32-20.76-55.47-51.99-55.47-154.29 0-77.7-54.48-139.9-127.94-155.16V32c0-17.67-14.32-32-31.98-32s-31.98 14.33-31.98 32v20.84C118.56 68.1 64.08 130.3 64.08 208c0 102.3-36.15 133.53-55.47 154.29-6 6.45-8.66 14.16-8.61 21.71.11 16.4 12.98 32 32.1 32h383.8c19.12 0 32-15.6 32.1-32 .05-7.55-2.61-15.27-8.61-21.71z">
-    </path>
-</svg>
+<?xml version="1.0" encoding="UTF-8" standalone="no"?> <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M224 512c35.32 0 63.97-28.65 63.97-64H160.03c0 35.35 28.65 64 63.97 64zm215.39-149.71c-19.32-20.76-55.47-51.99-55.47-154.29 0-77.7-54.48-139.9-127.94-155.16V32c0-17.67-14.32-32-31.98-32s-31.98 14.33-31.98 32v20.84C118.56 68.1 64.08 130.3 64.08 208c0 102.3-36.15 133.53-55.47 154.29-6 6.45-8.66 14.16-8.61 21.71.11 16.4 12.98 32 32.1 32h383.8c19.12 0 32-15.6 32.1-32 .05-7.55-2.61-15.27-8.61-21.71z"/></svg>
```
Comment (atb00ker) [inline openwisp_notifications/static/openwisp-notifications/images/icons/icon-bell.svg:1]: Blank space after the last line is missing! :smile:

File: openwisp_notifications/static/openwisp-notifications/js/notifications.js [hunk 2]
```diff
@@ -350,7 +352,7 @@ function initWebSockets($) {
             toast.slideDown('slow', function () {
                 setTimeout(function () {
                     toast.slideUp('slow', function () {
-                        toast.remove();
+                        // toast.remove();
```
Comment (atb00ker) [inline openwisp_notifications/static/openwisp-notifications/js/notifications.js:355]: What is this remove?

**Review by atb00ker — COMMENTED** [review]
I have not tested, but some comments:
1. `[chores] Minnor changes` -> `[chores] Minor changes`
2. Do we want to merge this to master or create a branch? @nemesisdesign

File: openwisp_notifications/static/openwisp-notifications/js/notifications.js (hunk 2 above)
Comment (niteshsinha17) [reply openwisp_notifications/static/openwisp-notifications/js/notifications.js:355]: It was not ready that why I was not moving to it ``ready`` column :smile:  And I commented this to restyle toast message. I will uncomment it when it gets ready.

>Do we want to merge this to master or create a branch? @nemesisdesign

I had created one branch named ``new-menu`` like we were making in other modules.

**Comment by atb00ker** [comment]
Hey @niteshsinha17 commit messages are not fixed as per the last review, do you need help with it?

**Comment by niteshsinha17** [comment]
> Hey @niteshsinha17 commit messages are not fixed as per the last review, do you need help with it?

Thanks, I have changed them. :smile: I thought you were saying about the typo that I should not repeat.

File: openwisp_notifications/static/openwisp-notifications/js/object-notifications.js [hunk 3]
```diff
@@ -15,9 +15,9 @@
 
 function getObjectNotificationComponent() {
     return `
-    <div class="ow-object-notification-container">
+    <li class="ow-object-notification-container">
         <button id="ow-object-notify" class="ow-object-notify button" title="You are receiving notifications for this object.">
-            <div class="ow-icon ow-object-notify-bell"></div>
+            <i class="ow-icon ow-object-notify-bell"></i>
```
Comment (atb00ker) [inline openwisp_notifications/static/openwisp-notifications/js/object-notifications.js:20]: I think `<span>` is more appropriate then `<i>` here! 😄

File: requirements.txt [hunk 4]
```diff
@@ -1,8 +1,9 @@
 django-notifications-hq~=1.6.0
 channels~=2.4.0
 # TODO set the next point release of openwisp-users when ready
-openwisp-users @ https://github.com/openwisp/openwisp-users/tarball/master
-openwisp-utils[rest]
+openwisp-users @ https://github.com/openwisp/openwisp-users/tarball/issues/253-register-menu-groups
+# TODO set the next point release of openwisp-utils when ready
+openwisp-utils[rest] @ https://github.com/niteshsinha17/openwisp-utils/tarball/gsoc/new-navigation-menu-design
```
Comment (atb00ker) [inline requirements.txt:6]: Before merge, please make it point to `openwisp/openwisp-utils` from `niteshsinha17/openwisp-utils`.

**Review by atb00ker — COMMENTED** [review]
Too many unrelated changes, I think some of them could have been seperately taken care. 😄

File: openwisp_notifications/static/openwisp-notifications/js/object-notifications.js (hunk 3 above)
Comment (niteshsinha17) [reply openwisp_notifications/static/openwisp-notifications/js/object-notifications.js:20]: I thought we are using ``<i>`` for icons in other places so I changed it

Comment (atb00ker) [reply openwisp_notifications/static/openwisp-notifications/js/object-notifications.js:20]: `<i>` is actually wrong, we should change at other places, but it's not a priority.

**Review by atb00ker — COMMENTED** [review]
I am confused about what should I test here:

Design:
designed-btn

What I see: 
btn-dont-match

There is a significant UI & it's feel difference, honestly, I don't mind either but am I even checking the correct thing?

1. The buttons in the currently visible section is too large, can we make them smaller please?
2. Can we hide these buttons when there are no new notifications in the new design?

**Comment by niteshsinha17** [comment]
@atb00ker Please don't judge it by comparing it with design because in miro we have some limitations. We can improve it during the implemetation.

>The buttons in the currently visible section is too large, can we make them smaller please?

:+1: 

>Can we hide these buttons when there are no new notifications in the new design?

It was already implemented in that way. Should we change it?

**Comment by atb00ker** [comment]
> It was already implemented in that way. Should we change it?

I think it wasn't implemented before, but I think it makes sense in the new UI! :smile:
We can create a seperate issue for it and tackle on low priority later on, sounds good? 😄

**Comment by niteshsinha17** [comment]
> > It was already implemented in that way. Should we change it?
> 
> I think it wasn't implemented before, but I think it makes sense in the new UI! 
> We can create a seperate issue for it and tackle on low priority later on, sounds good?

Yes, It's a good idea. I will open an issue for it.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Please ensure z-index is correctly set, otherwise map tiles can cover the notification widget:

Screenshot from 2021-07-03 16-46-30

File: openwisp_notifications/templates/admin/base_site.html [hunk 5]
```diff
@@ -15,18 +15,17 @@
...
+  <div class="ow-notifications toggle-btn" id="openwisp_notifications">
+    <button id="ow-notification-btn"></button>
     {% unread_notifications %}
   </div>
   {% csrf_token %}
   {% notification_toast %}
   {% notification_widget %}
-{% endblock %}
-
+{% endblock  %}
```
Comment (atb00ker) [inline openwisp_notifications/templates/admin/base_site.html:27]: Change not required.
P.S If you haven't already, install linters and if you use vscode, add:
```
    "files.insertFinalNewline": true,
    "files.trimTrailingWhitespace": true,
```

to ensure file ends with a new line and trailing whitespaces are removed!

**Review by atb00ker — COMMENTED** [review]
The "mark all as read" and "Show unread only" options look like they are borrowed from the UI. Is there any reason why entire UI isn't like other buttons? (they are more round and turn a different shade of grey when in focus).
But it's going in the correct direction! :+1:

**Review by nemesisdesign — COMMENTED** [review]
Looks like it's on the right track. I've added some minor improvements in https://github.com/openwisp/openwisp-notifications/pull/184/commits/68df5335490d1be250cbdca3958bb0e097ad0f2d.
Will be doing more review and testing in the coming days.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Something odd I noticed when testing on openwisp-monitoring, I see 2 "send command" buttons in the device detail page, it doesn't happen on master, can you please double check?

File: openwisp_notifications/templates/admin/base_site.html (hunk 5 above)
Comment (niteshsinha17) [reply openwisp_notifications/templates/admin/base_site.html:27]: Thank you :smile:

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
### Notification icon

Could you change the notification widget icon to use the same technique of masking so we can change its color on hover/focus please?

**Review by nemesisdesign — APPROVED** [review]

**Pull request merged**